### PR TITLE
Refactor sharewithme class

### PR DESCRIPTION
**Pull request by amrita-shrestha**
This ocis-pr 8322 bought changes mentioned below which need refactoring on sharewithme class. 
```
graph/sharedWithMe: return shares for the same resource as a single driveItem

- multiple shares for the same resource are now returned as a single driveItem
- the id for that driveItem is for now based on the resourceId of the shared item
      {sharesstorageproviderid}${sharejailid}!{resourceid of shared item}
- each share is exposed as a separate permission on the remoteId
- the permission now has an invitation property which provides the id of the creator
      of the share
- the client.synchronize flag is now exposed on the top-level driveitem. If at
      least on share of a resource is in accepted state the client.synchronize flag
      will be set to true.
- the UI.Hidden flag is now exposed on the top-level driveitem. If at least on
      share of a resource is marked as hidden the UI.Hidden flag will be set to
      true.
 - the 'shared' property is no longer available (the relevant information from that
      moved to the 'invitation' property of the individual permissions.
```
DriveItemId not being similar with propfind fileid is fixed by this ocis-pr 8467

Related issue:
- https://github.com/owncloud/ocis-php-sdk/issues/204

### CODE REVIEW

**Comment by amrita-shrestha** [comment]
blocked by https://github.com/owncloud/ocis/issues/8331

**Comment by amrita-shrestha** [comment]
> blocked by [owncloud/ocis#8331](https://github.com/owncloud/ocis/issues/8331)

add workaround for this issue using constant role id 

checked with moodle-repository-ocis ✅

File: src/ShareReceived.php [hunk 1]
```diff
@@ -59,89 +54,38 @@ public function getName(): string
...
-        if (empty($parentReference->getDriveId()) && $this->isClientSynchronized() === false) {
-            return  null;
-        }
-        if (empty($parentReference->getDriveId()) && $this->isClientSynchronized() === true) {
+        $time = $this->shareReceived->getLastModifiedDateTime();
+        if (empty($time)) {
             throw new InvalidResponseException(
-                "Invalid driveId returned in parentReference of received share '" .
-                print_r($parentReference->getDriveId(), true) . "'"
+                "Invalid shared modified DateTime'" . print_r($time, true) . "'"
```
Comment (individual-it) [inline src/ShareReceived.php:75]:
```suggestion
                "Invalid last modified DateTime'" . print_r($time, true) . "'"
```

File: src/ShareReceived.php [hunk 2]
```diff
@@ -159,136 +103,78 @@ public function getRemoteItemId(): string
...
+                "Invalid share createdBy information '" . print_r($this->shareReceived->getCreatedBy(), true) . "'"
+            ) : $this->shareReceived->getCreatedBy()->getUser();
     }
 
     /**
-     * @return \DateTimeImmutable
      * @throws InvalidResponseException
      */
-    public function getRemoteItemSharedDateTime(): \DateTimeImmutable
+    public function getCreatedByUserName(): string
```
Comment (individual-it) [inline src/ShareReceived.php:121]:
```suggestion
    public function getCreatedByDisplayName(): string
```

File: src/ShareReceived.php [hunk 3]
```diff
@@ -159,136 +103,78 @@ public function getRemoteItemId(): string
...
 
     /**
-     * gets the first permission of the remote item
-     * in theory there might be more that one permission, but currently there is no such case in ocis
-     * @return Permission
+     * @return Permission[]
      * @throws InvalidResponseException
      */
-    private function getRemoteItemPermission()
+    public function getRemoteItemPermission()
```
Comment (individual-it) [inline src/ShareReceived.php:146]: this function seems only be used in tests, do we really need it?

**Review by individual-it — CHANGES_REQUESTED** [review]

**Comment by amrita-shrestha** [comment]
currently waiting for 

changes on accept/decline share 
- ocis/pull/7885

**Review by individual-it — APPROVED** [review]

**Pull request merged**